pkg/runtime/core/registry.go

```diff
@@ -24,11 +24,20 @@ import (
 	"github.com/kubeflow/trainer/pkg/runtime"
 )
 
-type Registry map[string]func(ctx context.Context, client client.Client, indexer client.FieldIndexer) (runtime.Runtime, error)
+type Registry map[string]RuntimeRegistrar
+type RuntimeRegistrar struct {
+	factory      func(ctx context.Context, client client.Client, indexer client.FieldIndexer) (runtime.Runtime, error)
+	dependencies []string
+}
 
 func NewRuntimeRegistry() Registry {
 	return Registry{
-		TrainingRuntimeGroupKind:        NewTrainingRuntime,
-		ClusterTrainingRuntimeGroupKind: NewClusterTrainingRuntime,
+		TrainingRuntimeGroupKind: RuntimeRegistrar{
+			factory: NewTrainingRuntime,
+		},
+		ClusterTrainingRuntimeGroupKind: RuntimeRegistrar{
+			factory:      NewClusterTrainingRuntime,
+			dependencies: []string{TrainingRuntimeGroupKind},
```

Comment:
Do you think that in the future users can introduce runtimes that have multiple dependencies ?

Reply:
One potential thing is chained resources, something like ClusterXXXBlueprint -> ClusterTrainingRuntime -> TrainingRuntime

Reply:
There are plugin mechanisms in scheduler-plugings, but this Runtime factory (pipeline) mechanism is our original concept.